compiler/fir/resolve/src/org/jetbrains/kotlin/fir/scopes/impl/FirAbstractImportingScope.kt

```diff
@@ -6,85 +6,128 @@
 package org.jetbrains.kotlin.fir.scopes.impl
 
 import org.jetbrains.kotlin.descriptors.ClassKind
+import org.jetbrains.kotlin.descriptors.Visibilities
 import org.jetbrains.kotlin.fir.FirSession
+import org.jetbrains.kotlin.fir.declarations.FirMemberDeclaration
 import org.jetbrains.kotlin.fir.declarations.FirResolvedImport
 import org.jetbrains.kotlin.fir.declarations.expandedConeType
+import org.jetbrains.kotlin.fir.moduleVisibilityChecker
 import org.jetbrains.kotlin.fir.resolve.ScopeSession
 import org.jetbrains.kotlin.fir.resolve.toSymbol
 import org.jetbrains.kotlin.fir.resolve.transformers.ensureResolvedForCalls
 import org.jetbrains.kotlin.fir.scopes.FirScope
 import org.jetbrains.kotlin.fir.scopes.unsubstitutedScope
 import org.jetbrains.kotlin.fir.symbols.impl.*
 import org.jetbrains.kotlin.name.ClassId
+import org.jetbrains.kotlin.name.FqName
 import org.jetbrains.kotlin.name.Name
 
+enum class FirImportingScopeFilter {
+    ALL, INVISIBLE_CLASSES, MEMBERS_AND_VISIBLE_CLASSES;
+
+    fun check(symbol: FirClassLikeSymbol<*>, session: FirSession): Boolean {
+        if (this == ALL) return true
+        // TODO: also check DeprecationLevel.HIDDEN and required Kotlin version
```

Comment:
`fir.deprecationStatus() == DeprecationLevel.HIDDEN` with #4120

Reply:
Ok, it's not so simple because visitors resolve types in source order but `deprecationStatus` needs annotations to be resolved, which means visiting declarations in topological order.

Also, this is a paradox and it crashes the old frontend:
```
// FILE: aIsDeprecated.kt
package aIsDeprecated

typealias A = Deprecated

// FILE: bIsNotDeprecated.kt
package bIsNotDeprecated

annotation class B(val message: String, val level: DeprecationLevel)

// FILE: a.kt
package x

import bIsNotDeprecated.*

@B("", level = DeprecationLevel.HIDDEN)
annotation class A(val message: String, val level: DeprecationLevel)

// FILE: b.kt
package x

import aIsDeprecated.*

@A("", level = DeprecationLevel.HIDDEN)
typealias B = Deprecated
```